src/DynamoCore/PackageManager/PackageDirectoryBuilder.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Dynamo.Core;
+using Dynamo.Models;
+using Dynamo.Utilities;
+using Greg.Requests;
+using Greg.Responses;
+using RestSharp.Serializers;
+
+namespace Dynamo.PackageManager
+{
+    /// <summary>
+    ///     This class provides all of the tools to build a Package directory of the correct structure.
+    /// </summary>
+    internal class PackageDirectoryBuilder
+    {
+        private readonly IFileSystem fileSystem;
+        private readonly IPathRemapper pathRemapper;
+
+        /// <summary>
+        ///     The class constructor
+        /// </summary>
+        /// <param name="fileSystem">For moving files around</param>
+        /// <param name="pathRemapper">For modifying custom node paths</param>
+        internal PackageDirectoryBuilder(IFileSystem fileSystem, IPathRemapper pathRemapper)
+        {
+            if (fileSystem == null) throw new ArgumentNullException("fileSystem");
+            if (pathRemapper == null) throw new ArgumentNullException("pathRemapper");
+
+            this.fileSystem = fileSystem;
+            this.pathRemapper = pathRemapper;
+        }
+
+        #region Utility methods
+
+        /// <summary>
+        ///     Forms a properly formed package directory
+        /// </summary>
+        /// <param name="package">The package to be formed</param>
+        /// <param name="targetDirectory">The directory where the package should end up.  It is
+        ///     assumed that this directory is empty.  If it doesn't exist,
+        ///     it will be created.
+        /// </param>
+        /// <param name="files">The collection of files to be moved</param>
+        /// <returns></returns>
+        internal IDirectoryInfo BuildDirectory(Package package, string packagesDirectory, IEnumerable<string> files)
+        {
```

Comment:
All of the code here is basically copy pasted from `PackageUploadBuilder.cs` and thoroughly mocked.